README.md
# tianma-mount

![build status](https://travis-ci.org/tianmajs/tianma-static.svg?branch=master)

提供静态文件服务。

## 安装

    $ npm install tianma-static

## 使用

### 根目录

默认配置下，使用工作目录作为根目录。

    var tianma = require('tianma');

    tianma(8080)
        .static();

可以使用简洁配置指定根目录。

    var tianma = require('tianma');

    tianma(8080)
        .static('./htdocs');

也可以使用完整配置指定根目录。

    var tianma = require('tianma');

    tianma(8080)
        .static({ root: './htdocs' });

### 目录列表

当请求的是一个目录时，默认返回目录下的文件列表。可通过以下方式禁用该功能。

    var tianma = require('tianma');

    tianma(8080)
        .static({ root: './htdocs', indexes: false });

### 默认文件

当请求的是一个目录时，请求可以被自动重定向到默认的文件上。不存在默认文件时，则返回目录下的文件列表。

    var tianma = require('tianma');

    tianma(8080)
        .static({
            root: './htdocs',
            indexes: [ 'index.html', 'default.html' ]
        });

## 授权协议

[MIT](https://github.com/tianmajs/tianmajs.github.io/blob/master/LICENSE)


HISTORY.md
## 1.0.0 / 2015-03-10

+ 首次发布

## 1.0.1 / 2015-07-24

+ 重构代码
+ 简化模块设计

## 1.0.2 / 2015-07-27

+ 放弃`let`，改用`var`，以规避低版本NodeJS中的bug。

## 1.0.3 / 2015-07-28

+ `1.0.2`发布到npm之后代码文件丢失，重新发布。



test/static-spec.js
'use strict';

var http = require('http');
var static_ = require('..');
var request = require('supertest');
var tianma = require('tianma');
var path = require('path');

var FIXTURES_DIR = path.join(__dirname, 'fixtures');

function createApp() {
    var app = tianma();
    var server = http.createServer(app.run);

    app.server = server;

    return app;
}

describe('static({ root: root })', function () {
    function createServer() {
        var app = createApp();

        app.use(static_({ root: FIXTURES_DIR }));

        return app.server;
    }

    it ('should support access a file', function (done) {
        request(createServer())
            .get('/index.html')
            .expect(200)
            .expect('content-type', 'text/html')
            .expect('last-modified', /GMT/)
            .expect('<!DOCTYPE html>\n')
            .end(done);
    });

    it ('should support access a directory', function (done) {
        request(createServer())
            .get('/css/')
            .expect(200)
            .expect('content-type', 'text/html')
            .expect('last-modified', /GMT/)
            .expect(/a\.css<\/a>/)
            .end(done);
    });

    it ('should return 404 while missing something', function (done) {
        request(createServer())
            .get('/assets/b.css')
            .expect(404)
            .end(done);
    });
});

describe('static({ root: root, indexes: arr })', function () {
    function createServer() {
        var app = createApp();

        app.use(static_({
            root: FIXTURES_DIR,
            indexes: [ 'index.html', 'a.css' ]
        }));

        return app.server;
    }

    it ('should use the 1st default index', function (done) {
        request(createServer())
            .get('/')
            .expect(200)
            .expect('content-type', 'text/html')
            .expect('last-modified', /GMT/)
            .expect('<!DOCTYPE html>\n')
            .end(done);
    });

    it ('should use the 2nd default index', function (done) {
        request(createServer())
            .get('/css/')
            .expect(200)
            .expect('content-type', 'text/css')
            .expect('last-modified', /GMT/)
            .expect('.a {}\n')
            .end(done);
    });

    it ('should fall back to auto index when no default index found', function (done) {
        request(createServer())
            .get('/js/')
            .expect(200)
            .expect('content-type', 'text/html')
            .expect('last-modified', /GMT/)
            .expect(/a\.js<\/a>/)
            .end(done);
    });
});

describe('static({ root: root, indexes: false })', function () {
    function createServer() {
        var app = createApp();

        app.use(static_({
            root: FIXTURES_DIR,
            indexes: false
        }));

        return app.server;
    }

    it ('should ignore a directory', function (done) {
        request(createServer())
            .get('/css')
            .expect(404)
            .end(done);
    });

    it ('should access a file normally', function (done) {
        request(createServer())
            .get('/css/a.css')
            .expect(200)
            .expect('content-type', 'text/css')
            .expect('last-modified', /GMT/)
            .expect('.a {}\n')
            .end(done);
    });
});

describe('static(root)', function () {
    function createServer() {
        var app = createApp();

        app.use(static_(FIXTURES_DIR));

        return app.server;
    }

    it('should support the simple config', function (done) {
        request(createServer())
            .get('/')
            .expect(200)
            .expect('content-type', 'text/html')
            .expect('last-modified', /GMT/)
            .expect(/index\.html<\/a>/)
            .end(done);
    });
});

describe('static()', function () {
    function createServer() {
        var app = createApp();

        app.use(static_());

        return app.server;
    }

    it('should use "./" as the default root', function (done) {
        request(createServer())
            .get('/')
            .expect(200)
            .expect('content-type', 'text/html')
            .expect('last-modified', /GMT/)
            .expect(/package\.json<\/a>/)
            .end(done);
    });
});




lib/tpl.js
'use strict';

var LRU = require('lru-cache'),
    hash = require('./hash');

var PATTERN_BACK_SLASH = /\\/g,
    PATTERN_LITERAL = /%>([\s\S]*?)<%/gm,
    PATTERN_QUOTE = /"/g,
    PATTERN_CR = /\r/g,
    PATTERN_LF = /\n/g,
    PATTERN_TAB = /\t/g,
    PATTERN_EXP = /<%=(.*?)%>/g;

var cache = LRU(1024);

/**
 * Compile a template.
 * @param tpl {string}
 * @return {Function}
 */
function compile(tpl) {
    return new Function(
        'data',
            'var __=[],write=function(){__.push.apply(__,arguments);};' +
            'with(data){' +
            ('%>' + tpl + '<%')
                .replace(PATTERN_LITERAL, function ($0, $1) {
                    return '%>'
                        + $1.replace(PATTERN_BACK_SLASH, '\\\\')
                            .replace(PATTERN_QUOTE, '\\"')
                            .replace(PATTERN_CR, '\\r')
                            .replace(PATTERN_LF, '\\n')
                            .replace(PATTERN_TAB, '\\t')
                        + '<%';
                })
                .replace(PATTERN_EXP, '",$1,"')
                .replace(PATTERN_LITERAL, '__.push("$1");') +
            '}return __.join("");'
    );
}

/**
 * Get compiled template, or render template with data.
 * @param tpl {string}
 * @param data {Object}
 * @return {string}
 */
module.exports = function (tpl, data) {
    var key = hash(tpl),
        fn = cache[key];

    if (!fn) {
        fn = cache[key] = compile(tpl)
    }

    return data ? fn(data) : fn;
};


lib/static.js
'use strict';

var fs = require('co-fs'),
    ofs = require('fs'),
    path = require('path'),
    util = require('util');

var TEMPLATE = ofs.readFileSync(path.join(__dirname, 'list.tpl'), 'utf-8');

/**
 * Get the pathname of the requested file.
 * @param root {string}
 * @param pathname {string}
 * @return {string}
 */
function resolve(root, base, pathname) {
    pathname = path.normalize(pathname);
    pathname = pathname.substring(base.length);
    pathname = path.join(root, pathname);
    return pathname < root ? root : pathname;
}

/**
 * Get the sorted directory list;
 * @param pathname {string}
 * @param dir {string}
 * @param indexes {Array}
 * @return {string}
 */
function* list(pathname, dir, indexes) {
    var names = yield fs.readdir(dir),
        d = [],
        f = [],
        index;

    if ((index = findIndex(names, indexes))) {
        return yield fs.readFile(path.join(dir, index));
    } else {
        for (let i = 0, len = names.length; i < len; ++i) {
            let stats = yield fs.stat(path.join(dir, names[i]));
            if (stats.isDirectory()) {
                d.push(names[i] + '/');
            } else {
                f.push(names[i]);
            }
        }

        d.sort();
        f.sort();

        return util.format(TEMPLATE, pathname, d.concat(f).map(function (name) {
            return '<li><a href="' + encodeURI(name) + '">'
                + name
                + '</a></li>';
        }).join('\n'));
    }
}

/**
 * Find a possible index file among the given filenames.
 * @param filenames {Array}
 * @pararm indexes {Array}
 * @return {string|null}
 */
function findIndex(filenames, indexes) {
    var len = indexes.length,
        i = 0,
        index;

    for (; i < len; ++i) {
        index = indexes[i];
        if (filenames.indexOf(index) !== -1) {
            return index;
        }
    }

    return null;
}

/**
 * Filter factory.
 * @param [config] {Object}
 * @return {Function}
 */
module.exports = function (config) {
    config = config || {};

    if (typeof config == 'string') {
        config = {
            root: config
        };
    }
    var root = config.root || './',
        autoIndex = typeof config.autoIndex == 'undefined' ? true : config.autoIndex,
        base = config.base || '',
        indexes = config.indexes || [];

    return function *(next) {
        var req = this.request,
            res = this.response,
            pathname = resolve(root, base, decodeURI(req.pathname));

        try {
            var stats = yield fs.stat(pathname);

            if (stats.isFile()) {
                res.status(200)
                    .type(path.extname(pathname))
                    .head('last-modified', stats.mtime.toGMTString())
                    .data(yield fs.readFile(pathname));
            }
            else if (stats.isDirectory()) {
                if (!autoIndex) {
                    res.status(403)
                        .data('Forbidden to access!');
                } else {
                    res.status(200)
                        .type('html')
                        .head('last-modified', stats.mtime.toGMTString())
                        .data(yield *list((req.base || '' + req.pathname), pathname, indexes));
                }
            } else {
                yield next;
            }
        } catch (err) {
            if (err.code === 'ENOENT') {
                res.status(404);
                yield next;
            } else {
                throw err;
            }
        }
    };
};

index.js
'use strict';

var fs = require('co-fs');
var path = require('path');
var util = require('util');

var TEMPLATE = [
    '<!DOCTYPE html>',
    '<meta charset="UTF-8"/>',
    '<title>%s</title>',
    '<style>',
    'body { font: 10pt monospace; }',
    'li { line-height: 160%; list-style: none; }',
    '</style>',
    '<h2>Index of %s</h2>',
    '<hr />',
    '<ul>',
    '%s',
    '</ul>'
].join('\n');

/**
 * Get the absolute pathname of the requested file.
 * @param root {string}
 * @param pathname {string}
 * @return {string}
 */
function resolve(root, pathname) {
    pathname = path.normalize(pathname);
    pathname = path.join(root, pathname);

    return pathname < root ? root : pathname;
}

/**
 * Join two pathnames and normalize the slash.
 * @param base {string}
 * @param pathname {string}
 * @return {string}
 */
function join(base, pathname) {
    return path.join(base, pathname).replace(/\\/g, '/');
}

/**
 * Get the sorted directory list.
 * @param base {string}
 * @param dir {string}
 * @param filenames {Array}
 * @return {string}
 */
function *list(base, dir, filenames) {
    var d = [ '..' ];
    var f = [];

    for (var i = 0, len = filenames.length; i < len; ++i) {
        var stats = yield fs.stat(path.join(dir, filenames[i]));
        if (stats.isDirectory()) {
            d.push(filenames[i] + '/');
        } else {
            f.push(filenames[i]);
        }
    }

    d.sort();
    f.sort();

    var list = d.concat(f).map(function (filename) {
        return '<li><a href="' + encodeURI(join(base, filename)) + '">'
            + filename
            + '</a></li>';
    }).join('\n');

    return util.format(TEMPLATE, base, base, list);
}

/**
 * Find a possible index file among the given filenames.
 * @param filenames {Array}
 * @pararm indexes {Array}
 * @return {string|null}
 */
function findIndex(filenames, indexes) {
    var len = indexes.length,
        i = 0,
        index;

    for (; i < len; ++i) {
        index = indexes[i];
        if (filenames.indexOf(index) !== -1) {
            return index;
        }
    }

    return null;
}

/**
 * Filter factory.
 * @param [config] {Object}
 * @return {Function}
 */
module.exports = function (config) {
    config = config || {};

    if (typeof config == 'string') {
        config = {
            root: config
        };
    }

    var root = config.root || './';
    var indexes = 'indexes' in config ? config.indexes : [];

    /**
     * Read something.
     * @param pathname {string}
     * @return {Object}
     */
    function *read(pathname) {
        var absolute = resolve(root, pathname);
        var stats = yield fs.stat(absolute);

        if (stats.isFile()) {
            return {
                pathname: pathname,
                stats: stats,
                data: yield fs.readFile(absolute)
            };
        } else if (indexes && stats.isDirectory()) {
            var filenames = yield fs.readdir(absolute);
            var index = findIndex(filenames, indexes);

            if (index) {
                return yield *read(join(pathname, index));
            } else {
                return {
                    pathname: pathname,
                    stats: stats,
                    data: yield *list(pathname, absolute, filenames)
                };
            }
        } else {
            var err = new Error();
            err.code = 'ENOENT';
            throw err;
        }
    }

    return function *(next) {
        var req = this.request;
        var res = this.response;

        try {
            var result = yield *read(decodeURI(req.pathname));

            res.status(200)
                .type(result.stats.isDirectory() ?
                    'html': path.extname(result.pathname) || 'txt')
                .head('last-modified',
                    result.stats.mtime.toGMTString())
                .data(result.data);
        } catch (err) {
            if (err.code === 'ENOENT') {
                res.status(404);
                yield next;
            } else {
                throw err;
            }
        }
    };
};
